tests/test-dladdr.c

```diff
@@ -0,0 +1,195 @@
+/* On Unix like os compile with "-Wl,--export-dynamic -fpie" (default with cmake) */
+
+/* required for non Windows builds, must be set in front of the first system include */
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int verbose = 0;
+
+typedef enum {
+    Pass = 1,
+    PassWithoutSymbol = 2,
+    PassWithDifferentAddress = 3,
+    Fail = 0,
+    NoInfo = -1,
+} ExpectedResult;
+
+typedef void (* func) (void);
+
+static void print_dl_info( Dl_info *info, char *prefix, char *suffix )
+{
+    printf( "%sfilename: %s base: %p symbol name: '%s' addr: %p%s", prefix, info->dli_fname, info->dli_fbase, info->dli_sname, info->dli_saddr, suffix );
+}
+
+/**
+ * @brief check information returned by dladdr
+ * @param addr address to check
+ * @param addrsym
+ * @param expected check against expected result
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr( void *addr, char *addrsym, ExpectedResult expected_result )
+{
+    Dl_info info;
+    int result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match && !info.dli_sname)
+                 || (expected_result == PassWithDifferentAddress && sym_match && !addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
```

Comment:
This check is also incorrect. When test is expected to fail then sym_match must be always unset. And not only conditionally.

Reply:
Not sure, that I understand that statement correctly: A failure could be caused by not having a symbol match or not having an address match, which can be coded as `!(sym_match && addr_match)` or as `!sym_match || !addr_match` (see http://tma.main.jp/logic/logic.php?lang=en&type=eq&eq=%7EA%2B%7EB and http://tma.main.jp/logic/logic.php?lang=en&type=eq&eq=%7EA%2B%7EB).  Is the issue the readability by choosing the first form or something else ?

Reply:
Sorry, this is my mistake, you are right, ignore my comment.